File: LeetCode/121. Best Time to Buy and Sell Stock/elixir/solution.exs
defmodule Solution do
  @spec max_profit(prices :: [integer]) :: integer
  def max_profit([h | t]) do
    t
    |> Enum.reduce({h, h, 0}, fn price, {buy, sell, max_prfit} ->
      cond do
        price < buy -> {price, price, max_prfit}
        price > sell -> {buy, price, max(price - buy, max_prfit)}
        true -> {buy, sell, max_prfit}
      end
    end)
    |> then(fn {_buy, _sell, max_prfit} -> max_prfit end)
  end
end

File: LeetCode/121. Best Time to Buy and Sell Stock/java/Solution.java
class Solution {
    public int maxProfit(int[] prices) {
        int min = prices[0];
        int max = prices[0];
        int maxProfit = 0;
        for (int i = 1; i < prices.length; i++) {
            if (max < prices[i]) {
                max = prices[i];
                maxProfit = Math.max(maxProfit, max - min);
            }
            if (min > prices[i]) {
                min = prices[i];
                max = prices[i];
            }
        }
        return maxProfit;
    }
}